core/domain/stats_services.py

```diff
@@ -264,21 +269,31 @@ def advance_version_of_exp_stats(
 
         return exp_stats
 
+    state_name_stats_mapping = {}
+    old_state_names = utils.compute_list_difference(
+        exp_stats.state_stats_mapping,
+        exp_versions_diff.deleted_state_names)
+
     # Handling state deletions.
-    for state_name in exp_versions_diff.deleted_state_names:
-        exp_stats.state_stats_mapping.pop(state_name)
+    for state_name in old_state_names:
+        state_name_stats_mapping[state_name] = (
+            exp_stats.state_stats_mapping[state_name].clone())
```

Comment:
I feel like this for loop is unnecessary, actually. The deleted state names would not exist in the keys of new_to_old_state_names anyway, right? So the keys of new_to_old_state_names already gives you what you want here, I think.

Is that right? If so, you can probably delete this for loop as well as the "if old_state_name not in" clause below.

Reply:
How creation of new stats works is, we copy over a prev version state stats (which may contain extra states that weren't operated on), then we add / delete / handle renames for the diff of the versions. So without this loop, the states that aren't present in the diff will not be present in the stats model and that is not ideal.

Reply:
Oh -- got it, I think I see what you mean. But in that case, then, could you just pre-populate the stuff for state names that are neither renamed nor deleted?

Reply:
Done